// @trpc/server/adapters/next/dist/trpc-server-adapters-next.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../../../dist/index-4fbf1837.cjs.dev.js');
require('../../../dist/router-dae9f131.cjs.dev.js');
require('events');
require('http');
require('url');
require('../../../dist/once-34ddee5a.cjs.dev.js');

/* eslint-disable @typescript-eslint/no-explicit-any */
function createNextApiHandler(opts) {
  return async (req, res) => {
    function getPath() {
      if (typeof req.query.trpc === 'string') {
        return req.query.trpc;
      }

      if (Array.isArray(req.query.trpc)) {
        return req.query.trpc.join('/');
      }

      return null;
    }

    const path = getPath();

    if (path === null) {
      const error = opts.router.getErrorShape({
        error: new index.HTTPError('Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?', {
          statusCode: 500,
          code: 'INTERNAL_SERVER_ERROR'
        }),
        type: 'unknown',
        ctx: undefined,
        path: undefined,
        input: undefined
      });
      const json = {
        ok: false,
        statusCode: 500,
        error
      };
      res.statusCode = json.statusCode;
      res.json(json);
      return;
    }

    await index.requestHandler({ ...opts,
      req,
      res,
      path
    });
  };
}

exports.createNextApiHandler = createNextApiHandler;
