// @trpc/react/dist/trpc-react.esm.js
import { createTRPCClient } from '@trpc/client';
export * from '@trpc/client';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _slicedToArray from '@babel/runtime/helpers/esm/slicedToArray';
import _toConsumableArray from '@babel/runtime/helpers/esm/toConsumableArray';
import React, { createContext, useCallback, useEffect, useRef, useMemo } from 'react';
import { useQuery, useMutation, hashQueryKey, useInfiniteQuery } from 'react-query';
import { g as getCacheKey, C as CACHE_KEY_QUERY, a as CACHE_KEY_INFINITE_QUERY, b as CACHE_KEY_LIVE_QUERY } from './getCacheKey-634010d9.esm.js';

var TRPCContext = /*#__PURE__*/createContext(null);

function createReactQueryHooks() {
  var Context = TRPCContext;

  function createClient(opts) {
    return createTRPCClient(opts);
  }

  function TRPCProvider(_ref) {
    var client = _ref.client,
        queryClient = _ref.queryClient,
        children = _ref.children,
        _ref$isPrepass = _ref.isPrepass,
        isPrepass = _ref$isPrepass === void 0 ? false : _ref$isPrepass;
    return /*#__PURE__*/React.createElement(Context.Provider, {
      value: {
        queryClient: queryClient,
        client: client,
        isPrepass: isPrepass,
        fetchQuery: useCallback(function (pathAndArgs, opts) {
          var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_QUERY);
          return queryClient.fetchQuery(cacheKey, function () {
            return client.query.apply(client, _toConsumableArray(pathAndArgs));
          }, opts);
        }, [client, queryClient]),
        fetchInfiniteQuery: useCallback(function (pathAndArgs, opts) {
          var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_INFINITE_QUERY);
          return queryClient.fetchInfiniteQuery(cacheKey, function () {
            return client.query.apply(client, _toConsumableArray(pathAndArgs));
          }, opts);
        }, [client, queryClient]),
        prefetchQuery: useCallback(function (pathAndArgs, opts) {
          var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_QUERY);
          return queryClient.prefetchQuery(cacheKey, function () {
            return client.query.apply(client, _toConsumableArray(pathAndArgs));
          }, opts);
        }, [client, queryClient]),
        prefetchInfiniteQuery: useCallback(function (pathAndArgs, opts) {
          var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_INFINITE_QUERY);
          return queryClient.prefetchInfiniteQuery(cacheKey, function () {
            return client.query.apply(client, _toConsumableArray(pathAndArgs));
          }, opts);
        }, [client, queryClient]),
        invalidateQuery: useCallback(function (pathAndArgs) {
          var cacheKey = getCacheKey(pathAndArgs);
          return queryClient.invalidateQueries(cacheKey);
        }, [queryClient]),
        cancelQuery: useCallback(function (pathAndArgs) {
          var cacheKey = getCacheKey(pathAndArgs);
          return queryClient.cancelQueries(cacheKey);
        }, [queryClient]),
        setQueryData: useCallback(function (pathAndArgs, output) {
          var cacheKey = getCacheKey(pathAndArgs);
          queryClient.setQueryData(cacheKey.concat([CACHE_KEY_QUERY]), output);
          queryClient.setQueryData(cacheKey.concat([CACHE_KEY_INFINITE_QUERY]), output);
        }, [queryClient])
      }
    }, children);
  }

  function useContext() {
    return React.useContext(Context);
  }

  function _useQuery(pathAndArgs, opts) {
    var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_QUERY);

    var _useContext = useContext(),
        client = _useContext.client,
        isPrepass = _useContext.isPrepass,
        queryClient = _useContext.queryClient,
        fetchQuery = _useContext.fetchQuery;

    if (typeof window === 'undefined' && isPrepass && (opts === null || opts === void 0 ? void 0 : opts.ssr) !== false && !queryClient.getQueryCache().find(cacheKey)) {
      fetchQuery(pathAndArgs);
    }

    var query = useQuery(cacheKey, function () {
      return client.query.apply(client, _toConsumableArray(pathAndArgs));
    }, opts);
    return query;
  }

  function _useMutation(path, opts) {
    var client = useContext().client;
    var hook = useMutation(function (input) {
      return client.mutation(path, input);
    }, opts);
    return hook;
  }
  /* istanbul ignore next */

  /**
   * ⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️
   *  **Experimental.** API might change without major version bump
   * ⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠
   */


  function useSubscription(pathAndArgs, opts) {
    var _opts$enabled;

    var enabled = (_opts$enabled = opts === null || opts === void 0 ? void 0 : opts.enabled) !== null && _opts$enabled !== void 0 ? _opts$enabled : true;
    var queryKey = hashQueryKey(pathAndArgs);
    var client = useContext().client;
    return useEffect(function () {
      if (!enabled) {
        return;
      }

      var stopped = false;

      var _pathAndArgs = _slicedToArray(pathAndArgs, 2),
          path = _pathAndArgs[0],
          input = _pathAndArgs[1];

      var promise = client.subscriptionOnce(path, input);
      promise.then(function (data) {
        if (stopped) {
          return;
        }

        (opts === null || opts === void 0 ? void 0 : opts.onBatch) && opts.onBatch(data);
      }).catch(function (err) {
        if (stopped) {
          return;
        }

        (opts === null || opts === void 0 ? void 0 : opts.onError) && opts.onError(err);
      });
      return function () {
        stopped = true;
        promise.cancel();
      }; // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [queryKey, enabled]);
  }
  /* istanbul ignore next */

  /**
   * ⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️
   *  **Experimental.** API might change without major version bump
   * ⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠
   */


  function useLiveQuery(pathAndArgs, opts) {
    var _pathAndArgs2 = _slicedToArray(pathAndArgs, 2),
        path = _pathAndArgs2[0],
        userInput = _pathAndArgs2[1];

    var currentCursor = useRef(null);
    var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_LIVE_QUERY);
    var client = useContext().client;
    var hook = useQuery(cacheKey, function () {
      return client.subscriptionOnce(path, _objectSpread(_objectSpread({}, userInput), {}, {
        cursor: currentCursor.current
      }));
    }, opts);
    var lastItem = useMemo(function () {
      var raw = hook.data;

      if (typeof raw === 'undefined') {
        return undefined;
      }

      var last = raw[raw.length - 1];
      return last;
    }, [hook.data]);
    var data = lastItem === null || lastItem === void 0 ? void 0 : lastItem.data;
    var lastCursor = lastItem === null || lastItem === void 0 ? void 0 : lastItem.cursor;
    useEffect(function () {
      currentCursor.current = lastCursor;
      hook.refetch(); // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [lastCursor]);
    return _objectSpread(_objectSpread({}, hook), {}, {
      data: data
    });
  }

  function _useInfiniteQuery(pathAndArgs, // FIXME: this typing is wrong but it works
  opts) {
    var _useContext2 = useContext(),
        client = _useContext2.client,
        isPrepass = _useContext2.isPrepass,
        fetchInfiniteQuery = _useContext2.fetchInfiniteQuery,
        queryClient = _useContext2.queryClient;

    var cacheKey = getCacheKey(pathAndArgs, CACHE_KEY_INFINITE_QUERY);

    var _pathAndArgs3 = _slicedToArray(pathAndArgs, 2),
        path = _pathAndArgs3[0],
        input = _pathAndArgs3[1];

    if (typeof window === 'undefined' && isPrepass && (opts === null || opts === void 0 ? void 0 : opts.ssr) !== false && !queryClient.getQueryCache().find(cacheKey)) {
      fetchInfiniteQuery(pathAndArgs);
    }

    var query = useInfiniteQuery(cacheKey, function (_ref2) {
      var pageParam = _ref2.pageParam;

      var actualInput = _objectSpread(_objectSpread({}, input), {}, {
        cursor: pageParam
      });

      return client.query(path, actualInput);
    }, opts);
    return query;
  }

  function useDehydratedState(client, trpcState) {
    var transformed = useMemo(function () {
      if (!trpcState) {
        return trpcState;
      }

      return client.runtime.transformer.deserialize(trpcState);
    }, [client, trpcState]);
    return transformed;
  }

  return {
    Provider: TRPCProvider,
    createClient: createClient,
    useContext: useContext,
    useQuery: _useQuery,
    useMutation: _useMutation,
    useSubscription: useSubscription,
    useLiveQuery: useLiveQuery,
    useDehydratedState: useDehydratedState,
    useInfiniteQuery: _useInfiniteQuery
  };
}

export { createReactQueryHooks };
