// @trpc/server/dist/trpc-server.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('./index-e5809b27.cjs.prod.js');
var router = require('./router-513d1286.cjs.prod.js');
var subscription = require('./subscription-0d037008.cjs.prod.js');
var http = require('http');
var url = require('url');
require('./once-7f92f88f.cjs.prod.js');
require('events');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var http__default = /*#__PURE__*/_interopDefault(http);
var url__default = /*#__PURE__*/_interopDefault(url);

/* eslint-disable @typescript-eslint/no-non-null-assertion */
function createHttpHandler(opts) {
  return async (req, res) => {
    const endpoint = url__default['default'].parse(req.url).pathname.substr(1);
    await index.requestHandler({ ...opts,
      req,
      res,
      path: endpoint
    });
  };
}
function createHttpServer(opts) {
  const handler = createHttpHandler(opts);
  const server = http__default['default'].createServer((req, res) => handler(req, res));
  return {
    server,

    listen(port) {
      server.listen(port);
      const actualPort = port === 0 ? server.address().port : port;
      return {
        port: actualPort
      };
    }

  };
}

exports.HTTPError = index.HTTPError;
exports.assertNotBrowser = index.assertNotBrowser;
exports.getStatusCodeFromError = index.getStatusCodeFromError;
exports.httpError = index.httpError;
exports.requestHandler = index.requestHandler;
exports.Router = router.Router;
exports.router = router.router;
exports.Subscription = subscription.Subscription;
exports.subscriptionPullFactory = subscription.subscriptionPullFactory;
exports.createHttpHandler = createHttpHandler;
exports.createHttpServer = createHttpServer;
