// @trpc/server/adapters/next/dist/trpc-server-adapters-next.esm.js
import { H as HTTPError, r as requestHandler } from '../../../dist/index-1a4e8243.esm.js';
import '../../../dist/router-5bb70c03.esm.js';
import 'events';
import 'http';
import 'url';
import '../../../dist/once-a5ff19f1.esm.js';

/* eslint-disable @typescript-eslint/no-explicit-any */
function createNextApiHandler(opts) {
  return async (req, res) => {
    function getPath() {
      if (typeof req.query.trpc === 'string') {
        return req.query.trpc;
      }

      if (Array.isArray(req.query.trpc)) {
        return req.query.trpc.join('/');
      }

      return null;
    }

    const path = getPath();

    if (path === null) {
      const error = opts.router.getErrorShape({
        error: new HTTPError('Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?', {
          statusCode: 500,
          code: 'INTERNAL_SERVER_ERROR'
        }),
        type: 'unknown',
        ctx: undefined,
        path: undefined,
        input: undefined
      });
      const json = {
        ok: false,
        statusCode: 500,
        error
      };
      res.statusCode = json.statusCode;
      res.json(json);
      return;
    }

    await requestHandler({ ...opts,
      req,
      res,
      path
    });
  };
}

export { createNextApiHandler };
